### PR TITLE
Initial implementation of 'push bundle'

### PR DESCRIPTION
**Pull request by pcoop**

### CODE REVIEW

File: __tests__/api/BundlePush/BundlePusher.test.ts [hunk 1]
```diff
@@ -0,0 +1,515 @@
...
+        const listSpy = jest.spyOn(List, "fileList").mockImplementationOnce(() => { throw new Error( "Injected List error" ); });
+        await runPushTestWithError("__tests__/__resources__/ExampleBundle01",  false, "Injected List error");
+
+        expect(zosMFSpy).toHaveBeenCalledTimes(1);
+        expect(sshSpy).toHaveBeenCalledTimes(1);
+        expect(shellSpy).toHaveBeenCalledTimes(1);
+        expect(listSpy).toHaveBeenCalledTimes(1);
+    });
+    it("should complain if remote bundledir list fails", async () => {
+        const zosMFSpy = jest.spyOn(ZosmfSession, "createBasicZosmfSession").mockImplementationOnce(() => ({}));
```
Comment (matthewpwilson) [inline __tests__/api/BundlePush/BundlePusher.test.ts]: Could some of these spies that are common to most of the tests be set up in the beforeEach()   Would just make individual tests a bit easier to read.

File: docs/README.md [hunk 2]
```diff
@@ -142,6 +146,91 @@ The above command will attempt to disable and discard a CICS Bundle named MYBUND
 
   `zowe cics-deploy undeploy bundle --help`
 
+<a name="pushing"></a>
+## Pushing a Bundle to a CPSM managed CICS environment
+
+Pushing a Bundle to CICS is a composite action that will upload a bundle from the current directory to zFS, then install and enable it in the target CICS environment. This process consists of several stages which are combined together as a simple atomic action. Advanced users may prefer to issue the sub-commands individually, perhaps adding customised site-specific automation, rather than using the `push` action.
+
```
Comment (matthewpwilson) [inline docs/README.md:153]: Is it really atomic i.e. do we attempt to backout previous steps if a later step fails?

File: docs/README.md [hunk 3]
```diff
@@ -142,6 +146,91 @@ The above command will attempt to disable and discard a CICS Bundle named MYBUND
...
+<a name="push_req"></a>
+### Requirements for using push
+
+The cics-deploy push command combines several activities into a single combined action. In order to use the `push` command you will need several Zowe profiles to have been created. You will need a `zosmf` profile and an `ssh` profile, you might also consider creating a `cics-deploy` profile. You will also need a CICS bundle to exist in the current working directory (for further information on bundle enabling a directory see [Generating a CICS Bundle](#generation)).
+
+<a name="zosmf_profile"></a>
+#### The zosmf profile
+
+The `cics-deploy push bundle` command interacts with the target z/OS environment using several different protocols, one of which is zOSMF. The zosmf profile in Zowe defines the connectivity parameters required to connect to a remote zOSMF server in the target z/OS environment. You will typically require the hostname of the target server, the port number at which the zOSMF service is running, and a suitable userid and password for the target server. Your userid will need to have been authorised to perform the various actions that will be needed in order to deploy the Bundle.
```
Comment (matthewpwilson) [inline docs/README.md]:
```suggestion
The `cics-deploy push bundle` command interacts with the target z/OS environment using several different protocols, one of which is z/OSMF. The zosmf profile in Zowe defines the connectivity parameters required to connect to a remote z/OSMF server in the target z/OS environment. You will typically require the hostname of the target server, the port number at which the z/OSMF server is running, and a suitable user ID and password for the target server. Your user ID will need to have been authorised to perform the various actions that will be needed in order to deploy the Bundle.
```

File: docs/README.md [hunk 4]
```diff
@@ -142,6 +146,91 @@ The above command will attempt to disable and discard a CICS Bundle named MYBUND
...
+
+If you have not already done so, start by creating a zosmf profile in Zowe CLI. If you only create a single such profile then it will be used as the default z/OS configuration for any further zowe interactions. For further assistance on creating a z/OSMF profile issue the following command:
+
+  `zowe profiles create zosmf-profile --help`
+
+<a name="ssh_profile"></a>
+#### The ssh profile
+
+The `cics-deploy push bundle` command may attempt to issue Unix commands over the SSH protocol. In order to do so it requires a profile to be defined for the remote SSH server. You will typically need to know the hostname of the target server, the port at which the SSH service is running, and a suitable userid and password for the target server. It is recommended that you use the same userid and host for running SSH as is used for zOSMF, failure to do so will result in undefined behaviour.
```
Comment (matthewpwilson) [inline docs/README.md]:
```suggestion
The `cics-deploy push bundle` command may attempt to issue Unix commands over the SSH protocol. In order to do so it requires a profile to be defined for the remote SSH server. You will typically need to know the hostname of the target server, the port at which the SSH service is running, and a suitable user ID and password for the target server. It is recommended that you use the same user ID and host to connect with SSH as is used for z/OSMF, failure to do so will result in undefined behaviour.
```

File: docs/README.md [hunk 5]
```diff
@@ -142,6 +146,91 @@ The above command will attempt to disable and discard a CICS Bundle named MYBUND
...
+<a name="push_command"></a>
+### The push command
+
+The `cics-deploy push bundle` command pushes a bundle from the working directory to a target CICS environment. It requires the use of a zosmf profile and an ssh profile in order to do so, use of a cics-deploy profile is also encouraged (for more information on these dependencies see [Requirements for Push](#push_req)). You can request assistance on using the push command by issuing the following command:
+
+  `zowe cics-deploy push bundle --help`
+
+If you have configured default profiles for zosmf, ssh and cics-deploy then only two further items are needed:
+
+1. An up to 8 character unique name to use for the Bundle resource in CICS.
```
Comment (matthewpwilson) [inline docs/README.md]:
```suggestion
1. A unique name (up to 8 characters) to use for the BUNDLE resource in CICS.
```

File: src/api/BundlePush/BundlePusher.ts [hunk 6]
```diff
@@ -24,45 +26,289 @@ import { BundleDeployer } from "../BundleDeploy/BundleDeployer";
...
+    const zosmfProfile = this.params.profiles.get("zosmf");
+
+    if (zosmfProfile === undefined) {
+      throw new Error("No zosmf profile found");
+    }
+    return ZosmfSession.createBasicZosmfSession(zosmfProfile);
+  }
+
+  private async createSshSession(): Promise<SshSession> {
+    // Create a zosMF session
```
Comment (matthewpwilson) [inline src/api/BundlePush/BundlePusher.ts]:
```suggestion
```

**Review by matthewpwilson — CHANGES_REQUESTED** [review]

File: docs/README.md (hunk 2 above)
Comment (pcoop) [reply docs/README.md:153]: I've removed the word atomic.

File: __tests__/api/BundlePush/BundlePusher.test.ts (hunk 1 above)
Comment (pcoop) [reply __tests__/api/BundlePush/BundlePusher.test.ts]: I've restructured the tests.

Comment (matthewpwilson) [reply __tests__/api/BundlePush/BundlePusher.test.ts]: Thanks, they're much clearer now.

**Review by matthewpwilson — DISMISSED** [review]

**Review by matthewpwilson — APPROVED** [review]

**Pull request merged**